### DISCUSSION TITLE
Question about VAE and Fooocus Upscaling

### DISCUSSION THREAD

**Discussion by ILikeToasters** [lllyasviel/Fooocus#2861] [Q&A] [answered]
I have been experimenting with some new models like Cheyenne and Mohawk which both offer a version without VAE and another with VAE baked in. When using normally either one produces the same exact results. But when upscaling 1.5x or 2x and using the one without the VAE baked in and especially with images generated with a higher guidance scale the resulting images are washed out. I have not run into this before and curious why it happens since that is part of Fooocus I do not fully understand. Why would it only do it when upscaling? As seen in the screenshot the generated upscale is faded compared to the original generation. No settings changed after generating and then upscaling. 
<img width="405" alt="Screenshot 2024-05-03 144635" src="https://github.com/lllyasviel/Fooocus/assets/4357657/3f0e2bff-89c8-4da9-808e-ecac06a5dcea">

**Comment by ILikeToasters** [comment]
@mashb1t wondering if you would know the answer to this.

**Reply by mashb1t** [reply]
As there (in general, not only for upscaling) is no way to set the VAE manually,one of these models needs to include the VAE:
- base model
- refiner model

See https://github.com/lllyasviel/Fooocus/blob/8ef1a2119c79de18c0e296088b2f2c1d6b6d36da/modules/async_worker.py#L615-L620 as well as https://github.com/lllyasviel/Fooocus/blob/57a01865b99e3334fc83da25adc48ab989d853ab/modules/default_pipeline.py#L302-L314

EDIT: the Gradio preview uses xlvaeapp.pth or vaeapp_sd15.pth depending on if SDXL or not. I'm actually not quite sure why upscaling doesn't apply this VAE, but we've had a similar case with an ork image before somewhere in the issues.

**Reply by ILikeToasters** [reply]
I am aware of not being able to set the VAE, but why would the one without the VAE work fine for generating the image originally and exactly the same as the one with the VAE? I might be not understanding fully on this subject. I originally figured I would need the one with the VAE but I tested both to see how it would work. I guess I thought that if there was no VAE it would use a default one.

**Reply by mashb1t** [reply]
found it: https://github.com/lllyasviel/Fooocus/issues/1840

**Reply by ILikeToasters** [reply]
Almost missed your edit and was a little confused. lol. But now it makes sense and was basically what I figured then. It was more a curiosity than anything else so thanks for answering.

**Reply by mashb1t** [reply]
I can reproduce the behavior with [Mohawk](https://civitai.com/models/144952?modelVersionId=286100) (no VAE baked in)

![image](https://github.com/lllyasviel/Fooocus/assets/9307310/bf284ed7-7a52-4f0e-b721-a2b96c7f333b)

Let me see what i can do.

**Reply by mashb1t** [reply, marked as answer]
@ILikeToasters i did a thing, implemented in https://github.com/lllyasviel/Fooocus/pull/2867
Would appreciate it if you could test the feature. Thanks!

**Reply by ILikeToasters** [reply]
Testing now and do see the dropdown to choose a VAE and shows "Default (model)". Is Fooocus supposed to automatically get the VAE for ones without or is the idea I upload the VAE into the VAE folder to choose? Just want to be sure of expected behavior for this since I know some things Fooocus will download if not already in folders like the default models.

**Reply by ILikeToasters** [reply]
It does work though. Just tested and results were good.

**Reply by mashb1t** [reply]
@ILikeToasters automated fallback is not included. The feature offers a complete override for the VAE with one you choose, so it exactly behaves as you'd expect without any magic going on. The default behavior also stays the same.
No extra download required, but you have the option to put your own files in the models/vae folder and choose one of them.

**Reply by ILikeToasters** [reply]
Then it works as expected! Nice work!

**Reply by mashb1t** [reply]
merged to develop
